File: src/julia/union_find.jl
export UnionFind, is_root, root, unify, new_node, num_nodes, num_clusters, reset

type UnionFind
  parents :: Vector{Int}
  nnodes :: Int
  nclusters :: Int
  UnionFind() = new(zeros(Int,0),0,0)
  UnionFind(n::Int) = new(reshape(1:n,n),n,n)
end

num_nodes(u::UnionFind) = u.nnodes
num_clusters(u::UnionFind) = u.nclusters

is_root(u::UnionFind, n::Int) = u.parents[n] == n

function root_and_weight(u::UnionFind, n::Int)
  r,w = n,0
  while !is_root(u,r)
    r = u.parents[r]
    w += 1
  end
  u.parents[n] = r
  return r,w
end

root(u::UnionFind, n::Int) = root_and_weight(u,n)[1]

function unify(u::UnionFind, n1::Int, n2::Int)
  r1,w1 = root_and_weight(u,n1)
  r2,w2 = root_and_weight(u,n2)
  if r1 != r2
    u.nclusters -= 1
    if w1<w2
      u.parents[r1] = r2
    else
      u.parents[r2] = r1
    end
  else
    return r1
  end
end

function new_node(u::UnionFind) 
  push!(u.parents,length(u.parents)+1)
  u.nclusters += 1
  u.nnodes += 1
end

File: python/unionfind.py
class UnionFind:
  def __init__(self, num=0):
    self.parents = range(num)
    self.weights = [1]*num
    self.cluster_IDs = None

  def reset(self, num=0):
    self.parents = range(num)
    self.weights = [1] * num
    self.cluster_IDs = None
    return self

  def num(self):
    return len(self.parents)
  
  def isroot(self, index):
    return self.parents[index] == index

  def root(self, index):
    while not(self.isroot(index)):
      index = self.parents[index]
    return index

  def add(self):
    self.cluster_IDs = None
    N = self.num()
    self.parents.append(N)
    self.weights.append(1)
    return N

  def unify(self, index0, index1):
    if index0 == index1 :
      return index0
    self.cluster_IDs = None
    rt0,rt1 = self.root(index0), self.root(index1)
    w0,w1 = self.weights[rt0], self.weights[rt1]
    if w0 < w1 :
      rt0,rt1 = rt1,rt0

    self.parents[rt1] = rt0
    if w0 == w1:
      self.weights[rt0] += 1
    return rt0

  def clusterize(self):
    N = 0
    self.cluster_IDs = range(self.num())
    for index in xrange(self.num()):
      if self.isroot(index):
        self.cluster_IDs[index] = N
        N += 1
    for index in xrange(self.num()):
      rt = self.root(index)
      self.cluster_IDs[index] = self.cluster_IDs[rt]
    return N

  def cluster_id(self, index):
    if self.cluster_IDs == None:
      self.clusterize()
    return self.cluster_IDs[index]
